### PR TITLE
Flexible keys

### PR DESCRIPTION
**Pull request by andytill**
Tests for https://github.com/basho/riak_kv/pull/1357,  https://github.com/basho/riak_ql/pull/108

### CODE REVIEW

**Comment by jonmeredith** [comment]
Nice start.  Interesting whether to do this as a standard r_t or start using the shell.  These comments are based on the approach in this PR

General comments
- [ ] Would prefer descriptive names in subtests.
- [ ] Put data generator closer to query definition/tests - have to go between.

Tests:

- [ ] Insufficient keys written.
     - [ ] Need to exercise all levels of the key. Queries don't span quanta.
     - [ ] Need to try more than single partkey retrieval.
  - [ ] Insufficient types exercised. Need to exercise all types.

Test cases I want to see

 - [x] additional fields in local key after partition key
 - [x] partition key equal to local key
 - [x] quanta not the last key
 - [ ] partition keys of length 1-5 (check for anything limited by 3)
 - [ ] local keys of length 1-7

Schema Edge cases
 - [x] quanta-only partition key - yes, table 2
 - [ ] multiple timestamp fields - before and after quanta field
 - [ ] duplicated names in key
 - [ ] NULL fields

Query cases
 - [ ] Timestamps span more than one quanta

Failure cases
 - [ ] WHERE clauses without equality comparison against partition keys.
 - [ ] multiple quanta fields.

**Comment by javajolt** [comment]
@andytill Why do we have change sets going back to October 2015?  Which branch did you start with?

**Comment by andytill** [comment]
@brett those changes came from https://github.com/basho/riak_test/pull/1013, you would need to ask @gordonguthrie.  However, I don't think it is a bad thing to bring our TS branch up to date with master.

**Comment by jonmeredith** [comment]
+1 on the flexible keys and the quantum test.

I did not review the CT changes, however I did verify they ran (which triggered the 301c407 commit) - @andytill , do you have stuff in your erlang code path others may not?

Thanks for filling out the test cases, it looks good and was well worth the effort.

All we need to do now is get some generative quickcheck testing working ;)

@javajolt we can merge this once the common_test enhancements are reviewed.

**Comment by andytill** [comment]
@jonmeredith no local changes or unpushed commits and don't get undef errors with tdiff, maybe it is the way I am running it that includes all deps?

    make quickbuild && ./riak_test -c rtdev -t tests/ts_cluster_keys_SUITE.erl

**Comment by jonmeredith** [comment]
If you fire up an erlang shell outside the TS sources, what does m(tdiff). show you?

**Comment by andytill** [comment]
Depends if I add it to the erlang path or not! A vanilla erl shell with no args gives an `undef` of course. Not sure why I didn't run into this problem.

**Comment by jonmeredith** [comment]
Any chance you manually symlinked it into your deps subdir in riak test
without modifying rebar.config?

File: rebar.config [hunk 1]
```diff
@@ -11,21 +11,22 @@
...
-        {druuid,       ".*",    {git, "git://github.com/kellymclaughlin/druuid.git",    {tag, "0.2"}}}
+        {lager,         ".*", {git, "git://github.com/basho/lager",                   {tag,    "2.0.3"}}},
+        {getopt,        ".*", {git, "git://github.com/jcomellas/getopt",              {tag,    "v0.4"}}},
+        {meck,          ".*", {git, "git://github.com/basho/meck.git",                {tag,    "0.8.2"}}},
+        {mapred_verify, ".*", {git, "git://github.com/basho/mapred_verify",           {branch, "master"}}},
+        {riakc,         ".*", {git, "git://github.com/basho/riak-erlang-client",      {branch, "riak_ts-develop"}}},
+        {riakhttpc,     ".*", {git, "git://github.com/basho/riak-erlang-http-client", {branch, "riak_ts-develop"}}},
+        {kvc,           "1.3.0", {git, "https://github.com/etrepum/kvc",              {tag,    "v1.3.0"}}},
+        {druuid,        ".*", {git, "git://github.com/kellymclaughlin/druuid.git",    {tag,    "0.2"}}},
+        {tdiff,         ".*", {git, "git://github.com/tomas-abrahamsson/tdiff",       {branch, "master"}}}
```
Comment (javajolt) [inline rebar.config]: Could we please `{tag, "0.1"}`?  This currently matches master

File: src/riak_test_escript.erl [hunk 2]
```diff
@@ -148,9 +148,12 @@ main(Args) ->
                     Tests0;
                 {_, undefined} ->
                     Tests0;
+                %% TODO don't understand what this does
```
Comment (javajolt) [inline src/riak_test_escript.erl]: Did you understand Jon's explanation?  If so, we can probably remove this comment.

File: src/riak_test_escript.erl [hunk 3]
```diff
@@ -282,14 +285,28 @@ filter_merge_meta(SMeta, CMeta, [Field|Rest]) ->
     end.
 
 %% Check for api compatibility
+is_runnable(TestSpec, {Tests, NotTests}) ->
+    case is_runnable_test(TestSpec) of
+        {Type, true}  -> {[{Type, TestSpec} | Tests], NotTests};
+        {Type, false} -> {[Tests,                     {Type, TestSpec} | NotTests]}
```
Comment (javajolt) [inline src/riak_test_escript.erl]: Whitespace?

File: .gitignore [hunk 4]
```diff
@@ -23,4 +23,5 @@ dialyzer_unhandled_warnings
 dialyzer_warnings
 .idea
 riak_test.iml
-
+ct_logs/*
+variables-riak_test@127.0.0.1
```
Comment (javajolt) [inline .gitignore]: Newline?

File: intercepts/intercept.erl [hunk 5]
```diff
@@ -66,9 +66,22 @@ add(Target, Intercept, Mapping, OutDir) ->
     ok = compile_and_load(Original, OrigAC, OutDir),
     ok = compile_and_load(Target, ProxyAC, OutDir).
 
+-spec add(module(), module(), mapping()) -> ok.
 add(Target, Intercept, Mapping) ->
     add(Target, Intercept, Mapping, undefined).
 
+%% @doc Cleanup proxy and backuped original module
```
Comment (javajolt) [inline intercepts/intercept.erl]: Is the purpose of this to set the node back to the initialisation state?

**Comment by javajolt** [comment]
Looks like it works for single tests, but struggles with batches of them
```
./riak_test -c rtdev -t ts_cluster_comprehensive -t basic_command_line
...
escript: exception error: bad argument
  in function  io:format/3
     called as io:format(<0.24.0>,"Tests to run: ~p~n",
                         [{basic_command_line,
                              [{id,-1},
                               {platform,<<"local">>},
                               {version,
                                   <<"riak_ts-1.2.0-11-gc9389e8-riak_ts-develop">>},
                               {project,<<"riak_ts">>}]},
                          {ts_cluster_comprehensive,
                              [{id,-1},
                               {platform,<<"local">>},
                               {version,
                                   <<"riak_ts-1.2.0-11-gc9389e8-riak_ts-develop">>},
                               {project,<<"riak_ts">>}]}])
  in call from riak_test_escript:main/1 (src/riak_test_escript.erl, line 169)
  in call from escript:run/2 (escript.erl, line 747)
  in call from escript:start/1 (escript.erl, line 277)
  in call from init:start_it/1
  in call from init:start_em/1
```

**Comment by andytill** [comment]
Moved to https://github.com/basho/riak_test/pull/1040

**Pull request closed without merging**